### ISSUE TITLE
Set up Testing Framework and Training Pipeline for ConsistencyCheckerAgent

### ISSUE THREAD

**Issue by Marwan-Adawi** [Marwan-Adawi/ABET-Accreditation-Automation-System#4]
### 1. Testing Framework Setup

First, create `tests/test_consistency_checker.py`:

```python
import pytest
import numpy as np
from agents.consistency_checker_agent import ConsistencyCheckerAgent
from agents.environments.consistency_checker_env import ConsistencyCheckEnvironment

class TestConsistencyCheckerAgent:
    @pytest.fixture
    def sample_documents(self):
        return {
            'course_reports': [{
                'course_id': 'CS101',
                'clos': [
                    {'id': 'CLO1', 'plos': ['PLO1', 'PLO2'], 'assessment_methods': ['exam', 'project']},
                    {'id': 'CLO2', 'plos': ['PLO2'], 'assessment_methods': ['assignment']}
                ]
            }],
            'clo_assessments': [{
                'clo_id': 'CLO1',
                'scores': [85, 92, 78],
                'assessment_method': 'exam'
            }],
            'syllabus': {
                'clos': [
                    {'id': 'CLO1', 'description': 'Understanding basic concepts'},
                    {'id': 'CLO2', 'description': 'Applying concepts to problems'}
                ]
            }
        }

    @pytest.fixture
    def env(self, sample_documents):
        return ConsistencyCheckEnvironment(sample_documents)

    @pytest.fixture
    def agent(self, env):
        return ConsistencyCheckerAgent(env)

    def test_environment_initialization(self, env):
        assert env.num_clos == 2
        assert env.num_plos == 2
        assert env.num_assessment_methods == 2

    def test_state_space(self, env):
        state, _ = env.reset()
        assert 'clo_plo_mapping' in state
        assert state['clo_plo_mapping'].shape == (2, 2)
        assert 'assessment_methods' in state
        assert 'scores_valid' in state
        assert 'syllabus_match' in state

    def test_action_space(self, env):
        action = {
            'clo_index': 0,
            'issue_type': 0,
            'flag': 1
        }
        state, reward, done, _, _ = env.step(action)
        assert isinstance(reward, float)
        assert isinstance(done, bool)

    def test_agent_training(self, agent):
        rewards = agent.train(num_episodes=5, return_rewards=True)
        assert len(rewards) == 5
        assert all(isinstance(r, float) for r in rewards)

    def test_model_prediction(self, agent, env):
        state, _ = env.reset()
        action = agent.select_action(state)
        assert all(k in action for k in ['clo_index', 'issue_type', 'flag'])
```

### 2. Training Visualization Tools

Create `utils/visualization.py`:

```python
import matplotlib.pyplot as plt
import numpy as np
from typing import List, Dict
import seaborn as sns
from datetime import datetime
import os

class TrainingVisualizer:
    def __init__(self, save_dir: str = "training_plots"):
        self.save_dir = save_dir
        os.makedirs(save_dir, exist_ok=True)
        
    def plot_training_rewards(self, rewards: List[float], 
                            window_size: int = 100,
                            show: bool = True,
                            save: bool = True):
        """Plot training rewards with moving average."""
        plt.figure(figsize=(12, 6))
        
        # Plot raw rewards
        plt.plot(rewards, alpha=0.3, label='Raw Rewards')
        
        # Plot moving average
        if len(rewards) >= window_size:
            moving_avg = np.convolve(rewards, 
                                   np.ones(window_size)/window_size, 
                                   mode='valid')
            plt.plot(range(window_size-1, len(rewards)), 
                    moving_avg, 
                    label=f'Moving Average (window={window_size})')
        
        plt.xlabel('Episode')
        plt.ylabel('Reward')
        plt.title('Training Rewards Over Time')
        plt.legend()
        
        if save:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            plt.savefig(f"{self.save_dir}/training_rewards_{timestamp}.png")
        
        if show:
            plt.show()
        else:
            plt.close()

    def plot_state_heatmap(self, state: Dict, show: bool = True, save: bool = True):
        """Plot heatmap of current state."""
        fig, axes = plt.subplots(2, 2, figsize=(15, 12))
        
        # Plot CLO-PLO mapping
        sns.heatmap(state['clo_plo_mapping'], 
                   ax=axes[0,0], 
                   cmap='YlOrRd',
                   xticklabels=[f'PLO{i+1}' for i in range(state['clo_plo_mapping'].shape[1])],
                   yticklabels=[f'CLO{i+1}' for i in range(state['clo_plo_mapping'].shape[0])])
        axes[0,0].set_title('CLO-PLO Mapping')
        
        # Plot Assessment Methods
        sns.heatmap(state['assessment_methods'], 
                   ax=axes[0,1],
                   cmap='YlOrRd',
                   xticklabels=[f'M{i+1}' for i in range(state['assessment_methods'].shape[1])],
                   yticklabels=[f'CLO{i+1}' for i in range(state['assessment_methods'].shape[0])])
        axes[0,1].set_title('Assessment Methods')
        
        # Plot Score Validity
        sns.heatmap(state['scores_valid'].reshape(-1, 1), 
                   ax=axes[1,0],
                   cmap='YlOrRd',
                   xticklabels=['Valid'],
                   yticklabels=[f'CLO{i+1}' for i in range(len(state['scores_valid']))])
        axes[1,0].set_title('Score Validity')
        
        # Plot Syllabus Match
        sns.heatmap(state['syllabus_match'].reshape(-1, 1),
                   ax=axes[1,1],
                   cmap='YlOrRd',
                   xticklabels=['Match'],
                   yticklabels=[f'CLO{i+1}' for i in range(len(state['syllabus_match']))])
        axes[1,1].set_title('Syllabus Match')
        
        plt.tight_layout()
        
        if save:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            plt.savefig(f"{self.save_dir}/state_heatmap_{timestamp}.png")
        
        if show:
            plt.show()
        else:
            plt.close()
```

### 3. Enhanced Training Script

Create `train_consistency_checker.py`:

```python
import torch
import numpy as np
from agents.consistency_checker_agent import ConsistencyCheckerAgent
from agents.environments.consistency_checker_env import ConsistencyCheckEnvironment
from utils.visualization import TrainingVisualizer
import json
import os
from datetime import datetime
import logging

def setup_logging():
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(levelname)s - %(message)s',
        handlers=[
            logging.FileHandler('training.log'),
            logging.StreamHandler()
        ]
    )

def load_sample_data(file_path: str = "sample_data.json"):
    with open(file_path, 'r') as f:
        return json.load(f)

def train_agent(documents: dict, 
                num_episodes: int = 1000,
                save_dir: str = "models",
                visualizer: TrainingVisualizer = None,
                eval_interval: int = 50):
    """Train the consistency checker agent with visualization and logging."""
    
    # Setup
    setup_logging()
    os.makedirs(save_dir, exist_ok=True)
    
    if visualizer is None:
        visualizer = TrainingVisualizer()
    
    # Create environment and agent
    env = ConsistencyCheckEnvironment(documents)
    agent = ConsistencyCheckerAgent(env)
    
    # Training tracking
    all_rewards = []
    best_avg_reward = float('-inf')
    
    # Training loop
    for episode in range(num_episodes):
        state, _ = env.reset()
        episode_reward = 0
        episode_steps = 0
        
        while True:
            # Select and perform action
            action = agent.select_action(state)
            next_state, reward, done, _, _ = env.step(action)
            episode_reward += reward
            episode_steps += 1
            
            # Store transition and optimize
            agent.memory.append((state, action, reward, next_state, done))
            agent._optimize_model()
            
            if done:
                break
                
            state = next_state
        
        # Track rewards
        all_rewards.append(episode_reward)
        
        # Log progress
        if (episode + 1) % 10 == 0:
            avg_reward = np.mean(all_rewards[-10:])
            logging.info(f"Episode {episode + 1}/{num_episodes} - "
                        f"Average Reward: {avg_reward:.2f}, "
                        f"Steps: {episode_steps}, "
                        f"Epsilon: {agent.epsilon:.3f}")
            
            # Visualize current state and rewards
            if (episode + 1) % 50 == 0:
                visualizer.plot_training_rewards(all_rewards)
                visualizer.plot_state_heatmap(state)
        
        # Save best model
        if (episode + 1) % eval_interval == 0:
            avg_reward = np.mean(all_rewards[-eval_interval:])
            if avg_reward > best_avg_reward:
                best_avg_reward = avg_reward
                model_path = os.path.join(save_dir, 
                                        f"best_model_reward_{avg_reward:.2f}.pt")
                torch.save({
                    'episode': episode,
                    'model_state_dict': agent.network.state_dict(),
                    'optimizer_state_dict': agent.optimizer.state_dict(),
                    'reward': avg_reward,
                }, model_path)
                logging.info(f"Saved best model with average reward: {avg_reward:.2f}")
    
    return agent, all_rewards

if __name__ == "__main__":
    # Load sample data
    documents = load_sample_data()
    
    # Create visualizer
    visualizer = TrainingVisualizer()
    
    # Train agent
    trained_agent, rewards = train_agent(
        documents=documents,
        num_episodes=1000,
        visualizer=visualizer
    )
    
    # Final visualization
    visualizer.plot_training_rewards(rewards)
```

### Usage Instructions

1. First, install required packages:
```bash
pip install pytest matplotlib seaborn torch numpy
```

2. Run tests:
```bash
pytest tests/test_consistency_checker.py -v
```

3. Train the agent:
```bash
python train_consistency_checker.py
```

### Next Steps

1. [ ] Create sample data file (sample_data.json) with realistic ABET documentation
2. [ ] Run initial training and adjust hyperparameters based on results
3. [ ] Add cross-validation to test suite
4. [ ] Implement early stopping based on validation performance
5. [ ] Add more visualization metrics (e.g., loss curves, action distributions)
6. [ ] Create evaluation script for trained models

### Training Monitoring

The training script provides:
- Real-time reward tracking
- State visualization
- Model checkpointing
- Comprehensive logging
- Performance metrics

Monitor the training progress through:
1. Console output with episode rewards and statistics
2. Training plots in the training_plots directory
3. Log file (training.log) with detailed information
4. Saved model checkpoints in the models directory
